libs/encryption-sdk/src/main/java/org/opensearch/encryption/frame/core/FrameDecryptionHandler.java

```diff
@@ -0,0 +1,296 @@
+        } else {
+            final ProcessingSummary result = new ProcessingSummary(actualOutLen, totalParsedBytes - unparsedBytes_.length);
+            unparsedBytes_ = new byte[0];
+            return result;
+        }
+    }
+
+    /**
+     * Finish processing of the bytes. This function does nothing since the final frame will be
+     * processed and decrypted in processBytes().
+     *
+     * @param out space for any resulting output data.
+     * @param outOff offset into out to start copying the data at.
+     * @return 0
+     */
+    @Override
+    public int doFinal(final byte[] out, final int outOff) {
+        if (!complete_) {
+            throw new BadCiphertextException("Unable to process entire ciphertext.");
+        }
+
+        return 0;
+    }
+
+    /**
+     * Return the size of the output buffer required for a processBytes plus a doFinal with an input
+     * of inLen bytes.
+     *
+     * @param inLen the length of the input.
+     * @return the space required to accommodate a call to processBytes and doFinal with len bytes of
+     *     input.
+     */
+    @Override
+    public int estimateOutputSize(final int inLen) {
+        int outSize = 0;
+
+        final int totalBytesToDecrypt = unparsedBytes_.length + inLen;
+        if (totalBytesToDecrypt > 0) {
+            int frames = totalBytesToDecrypt / frameSize_;
+            frames += 1; // add one for final frame which might be < frame size.
+            outSize += (frameSize_ * frames);
+        }
+
+        return outSize;
+    }
+
+    public static long estimateDecryptedSize(long encryptedSize, int frameSize, int nonceLen, int tagLenBytes) {
+        // Calculate the size of sequence number for the last frame
+        long lastFrameSeqNumberSize = (Integer.SIZE / Byte.SIZE);
+
+        // Calculate the size of the final frame size
+        long finalFrameSizeSize = (Integer.SIZE / Byte.SIZE);
+
+        // Calculate the total size of header overhead for the last frame
+        long lastFrameHeaderOverhead = lastFrameSeqNumberSize + finalFrameSizeSize;
+
+        // Calculate the number of frames
+        long frames = (encryptedSize - lastFrameHeaderOverhead) / (frameSize + nonceLen + tagLenBytes + (Integer.SIZE / Byte.SIZE)) + 1;
+
+        // Calculate the size of the actual content in frames
+        long contentSize = frames * frameSize;
+
+        // Calculate the size of the last frame
+        long lastFrameSize = 0;
+
+        // Calculate the sequence number size for all frames
+        long seqNumberSize = frames * (Integer.SIZE / Byte.SIZE);
+
+        // Calculate the total size of header overhead for all frames
+        long headerOverhead = (nonceLen + tagLenBytes) * frames + seqNumberSize;
+
+        // Calculate the size of the last frame content
+        lastFrameSize = encryptedSize - contentSize - headerOverhead - lastFrameHeaderOverhead;
+
+        return contentSize + lastFrameSize;
```

Comment:
the naming doesn't look correct . is `contentSize` not including the last frame size ? 

How can we find out the actual content size here  , given just the total enc size ? I mean for the same enc size , we can have different actual content sizes .

Reply:
If it is `estimatedSize` , how `testEncryptedDecryptedLengthEstimations` is working all fine and giving exact size  ?

Reply:
Yes, naming is not right. `contentSize` is just the size without last frame. Estimations in all places are not approximations. They are producing exact size of target content for a source content size.

I think you might had concern around using number of frames that why we were using `frames * frameSize` as the content size instead of `(frames - 1) * frameSize`. If yes, then you are right that it should have been frames - 1. Now, answer to question of why it was working is because eventual result was still correct. Let me explain how. 
C - Raw Content size excluding last frame
FL - Frame Length
LF - Last frame size (of raw content)  - For e.g. if C=105, FL = 100 then LF = 5.
With line `long lastFrameSize = encryptedSize - contentSize - headerOverhead - lastFrameHeaderOverhead`
lastFrameSize = Actual last frame Size i.e. LF - FL (Because contentSize contains an extra error frame)
With line `contentSize + lastFrameSize`
C + FL (Extra error frame) + LF - FL (previously calculated lastFrameSize)
This will cancel out extra FL and we will be left with C + LF which is the size we want.  And this is why tests were working :) .

I have fixed the bug in the latest commit.

We can always determine the decrypted size as long as we have encrypted size and encryption metadata.